code/modules/power/apc.dm

```diff
@@ -757,38 +757,55 @@
-		if(H.a_intent == INTENT_HARM)
-			if(eth_species.ethereal_charge >= ETHEREAL_CHARGE_FULL - 19)
-				to_chat(H, "<span class='warning'>Your charge is full!</span>")
+
+	if(isethereal(user))
+		var/mob/living/carbon/human/H = user
+		var/datum/species/ethereal/E = H.dna.species
+		if((H.a_intent == INTENT_HARM) && (E.drain_time < world.time))
+			if(cell.charge <= (cell.maxcharge / 2)) // if charge is under 50% you shouldn't drain it
+				to_chat(H, "<span class='warning'>The APC doesn't have much power, you probably shouldn't drain any.</span>")
 				return
-			if(cell.charge < 100)
-				to_chat(H, "<span class='warning'>The cell has no charge!</span>")
+			var/obj/item/organ/stomach/ethereal/stomach = H.getorganslot(ORGAN_SLOT_STOMACH)
+			if(!istype(stomach))
+				to_chat(H, "<span class='warning'>You can't receive charge!</span>")
+				return
+			if(stomach.crystal_charge >= ETHEREAL_CHARGE_FULL)
+				to_chat(H, "<span class='warning'>Your charge is full!</span>")
 				return
+			E.drain_time = world.time + 75
 			to_chat(H, "<span class='notice'>You start channeling some power through the APC into your body.</span>")
-			if(do_after(user, 30, target = src))
-				if(cell.charge >= 100 && eth_species.ethereal_charge < ETHEREAL_CHARGE_FULL - 19)
+			if(do_after(user, 75, target = src))
+				if(cell.charge <= (cell.maxcharge / 2) || (stomach.crystal_charge >= ETHEREAL_CHARGE_FULL))
+					to_chat(H, "<span class='warning'>You can't receive charge from the APC.</span>")
+					return
+				if(istype(stomach))
 					to_chat(H, "<span class='notice'>You receive some charge from the APC.</span>")
-					eth_species.adjust_charge(20)
-					cell.charge -= 50 //Charging is inefficient
+					stomach.adjust_charge(10)
+					cell.charge -= 10
 				else
 					to_chat(H, "<span class='warning'>You can't receive charge from the APC!</span>")
 			return
-		if(H.a_intent == INTENT_GRAB)
-			if(cell.charge >= cell.maxcharge)
+		if((H.a_intent == INTENT_GRAB) && (E.drain_time < world.time))
+			if(cell.charge == cell.maxcharge)
 				to_chat(H, "<span class='warning'>The APC is full!</span>")
 				return
-			if(eth_species.ethereal_charge < 20)
+			var/obj/item/organ/stomach/ethereal/stomach = H.getorganslot(ORGAN_SLOT_STOMACH)
+			if(!istype(stomach))
+				to_chat(H, "<span class='warning'>You can't transfer charge!</span>")
+				return
+			if(stomach.crystal_charge < 10)
 				to_chat(H, "<span class='warning'>Your charge is too low!</span>")
 				return
+			E.drain_time = world.time + 75
 			to_chat(H, "<span class='notice'>You start channeling power through your body into the APC.</span>")
-			if(do_after(user, 30, target = src))
-				if(cell.charge + 20 <= cell.maxcharge && (eth_species.ethereal_charge > 20))
+			if(do_after(user, 75, target = src))
+				if(cell.charge == cell.maxcharge || (stomach.crystal_charge < 10))
+					to_chat(H, "<span class='warning'>You can't transfer charge to the APC.</span>")
+					return
+				if(istype(stomach))

```

Comment:
This check should either be moved back, or removed completely since the stomach is already checked on line 793 and if the person somehow has their stomach changed while charging at an APC, line 802 will runtime from stomach not being the right type anyway